tools/reactive-controllers/src/RovingTabindex.ts

```diff
@@ -0,0 +1,304 @@
+        if (typeof direction === 'string') {
+            this._direction = () => direction;
+        } else if (typeof direction === 'function') {
+            this._direction = direction;
+        }
+        this.calculateFocusInIndex =
+            calculateFocusInIndex || this.calculateFocusInIndex;
+        this.elementEnterAction = elementEnterAction || this.elementEnterAction;
+        if (typeof listenerScope === 'object') {
+            this._listenerScope = () => listenerScope;
+        } else if (typeof listenerScope === 'function') {
+            this._listenerScope = listenerScope;
+        }
+    }
+
+    focus(options?: FocusOptions): void {
+        this.elements[this.currentIndex]?.focus(options);
+    }
+
+    private manageIndexesAnimationFrame = 0;
+
+    clearElementCache(): void {
+        delete this.cachedElements;
+        cancelAnimationFrame(this.manageIndexesAnimationFrame);
+        if (!this.managed) return;
+
+        this.manageIndexesAnimationFrame = requestAnimationFrame(() =>
+            this.manageTabindexes()
+        );
+    }
+
+    setCurrentIndexCircularly(diff: number): void {
+        const { length } = this.elements;
+        let nextIndex = (length + this.currentIndex + diff) % length;
+        while (
+            this.elements[nextIndex] &&
+            !this.isFocusableElement(this.elements[nextIndex])
+        ) {
+            nextIndex = (length + nextIndex + diff) % length;
+        }
+        this.currentIndex = nextIndex;
+    }
+
+    hostContainsFocus(): void {
+        this.host.addEventListener('focusout', this.handleFocusout);
+        this.host.addEventListener('keydown', this.handleKeydown);
+        this.focused = true;
+        const currentElement = this.elements[this.currentIndex];
+        if (!currentElement) return;
+    }
+
+    hostNoLongerContainsFocus(): void {
+        this.host.addEventListener('focusin', this.handleFocusin);
+        this.host.removeEventListener('focusout', this.handleFocusout);
+        this.host.removeEventListener('keydown', this.handleKeydown);
+        this.currentIndex = this.focusInIndex;
+        this.focused = false;
+        const currentElement = this.elements[this.currentIndex];
+        if (!currentElement) return;
+    }
+
+    isRelatedTargetAnElement(event: FocusEvent): boolean {
+        const relatedTarget = event.relatedTarget as null | Element;
+        return !this.elements.includes(relatedTarget as T);
+    }
+
+    handleFocusin = (event: FocusEvent): void => {
+        if (!this.isEventWithinListenerScope(event)) return;
+        if (this.isRelatedTargetAnElement(event)) {
+            this.hostContainsFocus();
+        }
+        const path = event.composedPath() as T[];
+        let targetIndex: number | undefined;
+        path.find((el) => {
+            targetIndex = this.elements.indexOf(el);
+            return targetIndex !== -1;
+        });
+        this.currentIndex = targetIndex ?? this.currentIndex;
```

Comment:
This may be doing exactly what you want as right now `targetIndex` can be `undefined`, but since indices are typically invalid at `-1` and valid at `0`, I'd recommend:

```suggestion
        this.currentIndex = targetIndex > -1 ? targetIndex : this.currentIndex;
```